Support Custom SidecarDB Names on VTTablets

mattlord opened the pull request:
## Description

This PR adds a new **experimental** feature: *keyspace level sidecar database names*. For the *why*, please see [the feature request](https://github.com/vitessio/vitess/issues/12304).

In this PR the sidecar db name can only be set when the keyspace is first created via the `CreateKeyspace` vtctl client command using the new `--sidecar-db-name` command flag. From that point forward it is immutable (w/o modifying the topo directly outside of Vitess). If no value is specified for the keyspace then it will be set to the default of `_vt` when the keyspace is created, and as a fallback the default will also be set when the first tablet is started within the keyspace if for any reason it's not already set in the topo keyspace record.

The management of the sidecar database resides entirely within a `vttablet`. This allows each `vttablet` in a keyspace to manage its own sidecar database independently of all other keyspaces and processes in the cluster. This allows e.g. for most tablets within a Vitess cluster to use the default sidecar name of `_vt` while some can use a non-standard name of e.g. `_vt_import` or `__virtual_vitess_cluster` for whatever purposes as needed (in various cases the tablet's `mysqld` instance may be participating in two distinct vitess clusters or two distinct keyspaces). From outside of the `vttablet`, Vitess is able to assume that the standard name of `_vt` is being used (including external/3rd party tooling that e.g. may be doing things like `vtctlclient ExecuteFetchAsDba`) unless talking directly to a tablet's mysqld instance.

>**Note** there are two exceptions to the above rule today — where another process *does* need to know the sidecar database name that a `vttablet` is using (by way of the keyspace it's in). Both are in `vtgate` and are due to the fact that today `vtgate` sends raw SQL queries to a `vttablet` through the normal user query hot path (tabletserver's `QueryExecutor.Execute()`) in the following cases:
> 1. OnlineDDL's `show vitess_migrations` implementation
> 2. VTGate's Schema Tracker implementation
>
> The plan is to eliminate these two exceptions in the future by leveraging new `vttablet` RPCs that `vtgate` can then use instead of the current SQL interface. This way we have proper encapsulation and `vtgate` does not need to be aware of the implementation details (in this case the schema `vttablet` uses for the storage).

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/12304

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Documentation: https://github.com/vitessio/website/pull/1402

shlomi-noach commented:
@mattlord will it make sense to have the sidecar db name _predictable_, possibly even as default,m without the need to add `--sidecar-db-name`?

e.g. is my keyspace is `commerce`, the sidecar DB name would be `_vt_commerce` ; in this case the tablet can first automatically search for sidecar db names as it starts; it can auto correlate the keyspace/sidecardb. It may choose to fall back to `_vt` if there's only one key space and no `_vt_<keyspace>` DB, etc.

mattlord commented:
> @mattlord will it make sense to have the sidecar db name _predictable_, possibly even as default, without the need to add `--sidecar-db-name`?
> 
> e.g. is my keyspace is `commerce`, the sidecar DB name would be `_vt_commerce` ; in this case the tablet can first automatically search for sidecar db names as it starts; it can auto correlate the keyspace/sidecardb. It may choose to fall back to `_vt` if there's only one key space and no `_vt_<keyspace>` DB, etc.

Just FYI, the `--sidecar-db-name` vttablet flag is gone. You now have to specify the non-default sidecar-db-name when creating the keyspace. 

This is an interesting idea and not one that I'm opposed to in any way. For me, if it's still variable then we still need to read it from the topo server and the PR is mostly the same.

shlomi-noach commented:
> Currently the sidecar db name can only be set

Just to clarify to myself, by "Currently" you mean "in this PR", right? I normally assume that "currently" relates to existing code in `main`.

shlomi-noach commented on go/vt/sidecardb/sidecardb.go:
```diff
@@ -364,35 +396,44 @@ func (si *schemaInit) ensureSchema(table *sidecarTable) error {
 	}
 	for _, q := range sidecarDBInitQueries {
-		db.AddQuery(q, result)
+		db.AddQuery(fmt.Sprintf(q, GetIdentifier()), result)
```
Continued - again I feel like this is not idiomatic; we should use `ParseAndBind` instead of `fmt.Sprintf`

shlomi-noach commented on go/vt/sqlparser/utils.go:
```diff
@@ -113,3 +113,52 @@ func NormalizeAlphabetically(query string) (normalized string, err error) {
+		return String(upd), nil
+	}
+	return query, nil
+}
```
:+1:  nice

shlomi-noach commented on go/vt/vtctl/vtctl.go:
```diff
@@ -1832,6 +1833,7 @@ func commandCreateKeyspace(ctx context.Context, wr *wrangler.Wrangler, subFlags
 	baseKeyspace := subFlags.String("base_keyspace", "", "Specifies the base keyspace for a snapshot keyspace")
 	timestampStr := subFlags.String("snapshot_time", "", "Specifies the snapshot time for this keyspace")
 	durabilityPolicy := subFlags.String("durability-policy", "none", "Type of durability to enforce for this keyspace. Default is none. Possible values include 'semi_sync' and others as dictated by registered plugins.")
+	sidecarDBName := subFlags.String("sidecar-db-name", sidecardb.DefaultName, "Vitess sidecar database used for internal metadata.")
```
We might want to consider the magic value `"auto"` to auto-generate a sidecar db name, e.g. if my keyspace is `commerce` then the auto generated name would be `_vt_commerce`.

shlomi-noach commented on go/vt/vtgate/schema/tracker.go:
```diff
@@ -259,7 +262,8 @@ func (t *Tracker) updatedTableSchema(th *discovery.TabletHealth) bool {
 	}
 	bv := map[string]*querypb.BindVariable{"tableNames": tables}
-	res, err := th.Conn.Execute(t.ctx, th.Target, mysql.FetchUpdatedTables, bv, 0, 0, nil)
+	res, err := th.Conn.Execute(t.ctx, th.Target, fmt.Sprintf(mysql.FetchUpdatedTables, "_vt"),
```
another use of `fmt.Sprintf()` (recurring comment, I'll stop here)

shlomi-noach commented on go/vt/vttablet/onlineddl/executor.go:
```diff
@@ -76,13 +77,13 @@ var (
+	fmt.Sprintf(`update %s.schema_migrations set migration_status='val1' where migration_uuid='val2' and mysql_schema='val3' and shard='val4'`, sidecardb.GetIdentifier()),
 }
 
 var vexecInsertTemplates = []string{
```
you may safely ignore this variable - it is unused and to be removed shortly.

shlomi-noach commented on go/vt/vttablet/onlineddl/executor.go:
```diff
@@ -3234,7 +3238,7 @@ func (e *Executor) runNextMigration(ctx context.Context) error {
 			return nil, err
 		}
-		r, err := e.execQuery(ctx, sqlSelectReadyMigrations)
+		r, err := e.execQuery(ctx, fmt.Sprintf(sqlSelectReadyMigrations, sidecardb.GetIdentifier()))
```
can we ensure the query always executes on the correct schema (ie the keyspace's sidecardb) rather than re-inject the schema name in all queries?

shlomi-noach reviewed (commented):
Wow, monumental work!

I think a recurring thing is that we now re-generate queries programmaticly in all places (I notice `onlineddl/executor.go` ; I feel this is cumbersome, can we instead ensure the connection uses the correct schema? We should aim to minimize the programmatic manipulations of queries, if possible. Maybe that's easier said than done, I'm not sure.

harshit-gangal commented on go/vt/vttablet/tabletserver/tabletserver.go:
```diff
@@ -730,6 +731,12 @@ func (tsv *TabletServer) Execute(ctx context.Context, target *querypb.Target, sq
 		return nil, vterrors.New(vtrpcpb.Code_INTERNAL, "[BUG] transactionID and reserveID must match if both are non-zero")
 	}
 
+	// Rewrite any sidecar database usage if needed.
+	sql, err = sqlparser.ReplaceTableQualifiers(sql, sidecardb.DefaultName, sidecardb.GetName())
+	if err != nil {
+		return nil, err
+	}
+
 	return tsv.execute(ctx, target, sql, bindVariables, transactionID, reservedID, nil, options)
```
this will happen for all incoming queries in the query serving path.

harshit-gangal commented on proto/topodata.proto:
```diff
@@ -304,6 +304,16 @@ message Keyspace {
   // DurabilityPolicy is the durability policy to be
   // used for the keyspace.
   string durability_policy = 8;
+
+  // ThrottlerConfig has the configuration for the tablet
+  // server's lag throttler, and applies to the entire
+  // keyspace, across all shards and tablets.
+  ThrottlerConfig throttler_config = 9;
+
+  // SidecarDBName is the name of the Vitess sidecar database
+  // used for various system metadata that is stored in each
+  // tablet's mysqld instance.
+  string sidecar_db_name = 10;
 }
```
should adding `ThrottlerConfig` be part of this PR?

shlomi-noach commented on go/vt/vttablet/onlineddl/executor.go:
```diff
@@ -284,7 +285,13 @@ func (e *Executor) execQuery(ctx context.Context, query string) (result *sqltype
+	if err != nil {
+		return nil, err
+	}
+	return conn.Exec(ctx, uq, math.MaxInt32, true)
```
This is a quick and powerful way to address the qualifier! I love it.

rohit-nayak-ps commented:
What about defining an interface with the required query Exec function signatures? We implement two concretes and at runtime select the correct one: 
1. sidecardb==`_vt`: will be a passthrough to the same function called today
1. different sidecardb name: this will first `ReplaceTableQualifiers()`

This way we don't incur any performance penalty for those using the recommended `_vt`. 

If `ReplaceTableQualifiers()` introduces unacceptable delays in the any path, we could see if introducing a cache where we store the original query and replaced query in a map helps

mattlord commented:
> What about defining an interface with the required query Exec function signatures? We implement two concretes and at runtime select the correct one:
> 
> 1. sidecardb==`_vt`: will be a passthrough to the same function called today
> 2. different sidecardb name: this will first `ReplaceTableQualifiers()`
> 
> This way we don't incur any performance penalty for those using the recommended `_vt`.
> 
> If `ReplaceTableQualifiers()` introduces unacceptable delays in the any path, we could see if introducing a cache where we store the original query and replaced query in a map helps

Good idea! I implemented this: https://github.com/vitessio/vitess/pull/12240/commits/c8cd425b64601a4a34a7d4e1816529d44ecd2735

Did you have something else in mind?

mattlord commented on go/vt/sidecardb/sidecardb.go:
```diff
@@ -103,7 +109,9 @@ func validateSchemaDefinition(name, schema string) (string, error) {
+// into a slice of sidecarTables for processing.
+func loadSchemaDefinitions() {
 	sqlFileExtension := ".sql"
 	err := fs.WalkDir(schemaLocation, ".", func(path string, entry fs.DirEntry, err error) error {
```
I don't disagree but it's also not related to this PR and I want to start limiting changes/refactoring I do that's not really part of this feature.

arthurschreiber commented:
Oh, this is super interesting to see! I think in `vtcombo`, this will allow us to have separate, per keyspace sidecar databases, instead of a single sidecar database that's incorrectly shared by all keyspaces. ❤️

shlomi-noach commented on go/vt/sqlparser/utils_test.go:
```diff
@@ -180,3 +181,85 @@ func TestQueryMatchesTemplates(t *testing.T) {
+			name: "simple set",
+			in:   "set names 'binary'",
+			out:  "set names 'binary'",
+		},
```
Can you please add tests with:
- a derived table 
- an inner select from another table

mattlord replied:
Done!

shlomi-noach commented on go/vt/vttablet/tabletserver/throttle/throttler.go:
```diff
@@ -401,6 +397,14 @@ func (throttler *Throttler) Open() error {
+	if throttleMetricQuery != "" {
+		throttler.metricsQuery.Store(throttleMetricQuery) // override
+	}
+	throttler.initConfig()
```
This looks valid :+1:

shlomi-noach commented on go/vt/vttablet/onlineddl/executor.go:
```diff
@@ -187,11 +188,14 @@ type Executor struct {
 
-	ticks             *timer.Timer
-	isOpen            int64
-	schemaInitialized bool
```
:+1:

shlomi-noach commented on go/vt/vttablet/onlineddl/executor.go:
```diff
@@ -187,11 +188,14 @@ type Executor struct {
+	ticks  *timer.Timer
+	isOpen int64
 
-	initVreplicationDDLOnce sync.Once
```
:+1:

shlomi-noach commented on go/vt/vttablet/tabletmanager/tm_init.go:
```diff
@@ -486,6 +487,34 @@ func (tm *TabletManager) createKeyspaceShard(ctx context.Context) (*topo.ShardIn
+		}
+		err = tm.TopoServer.UpdateKeyspace(ctx, ks)
+		if err != nil {
+			unlock(&lockErr)
+			return nil, vterrors.Wrap(err, "createKeyspaceShard: cannot GetOrCreateShard shard")
+		}
+		unlock(&lockErr)
```
looks like `unlock(&lockErr)` gets called no matter what. Perhaps write once, right after `UpdateKeyspace()`?

mattlord replied:
Good catch. Cleaned up.

shlomi-noach commented on go/vt/vttablet/tabletmanager/vreplication/controller_plan.go:
```diff
@@ -228,12 +254,23 @@ func buildDeletePlan(del *sqlparser.Delete) (*controllerPlan, error) {
 
 func buildSelectPlan(sel *sqlparser.Select) (*controllerPlan, error) {
-	switch sqlparser.ToString(sel.From) {
+	tableExpr, ok := sel.From[0].(*sqlparser.AliasedTableExpr)
```
I'm just thinking, could there be a scenario where `From` is an empty slice? e.g. `FROM DUAL` gets truncated? I know we won't be running `FROM DUAL` in `vreplication`; but I'd feel safer if we tested for the length of the slice.

mattlord replied:
Nice catch! Yes, will add some protection to this and any similar cases in these build functions.

shlomi-noach reviewed (approved):
This looks good, with a few comments left inline. I've reviewed throttler, online ddl, vreplication and general other stuff, but I definitely did not review query-serving related changes.

I know you've been working long on this PR and hope you can merge it soon!

mattlord commented on go/vt/sidecardb/identifier_cache.go:
```diff
@@ -0,0 +1,121 @@
+// NOT a singleton and allows you to override the global instance used
+// outside of unit tests.
+// NOTE: This should ONLY be used in unit tests and NOT in production!
+func NewTestIdentifierCache(loadFunc func(context.Context, string) (string, error)) *IdentifierCache {
```
I tried, but it doesn't work. Looks like go will not let you overwrite a pointer to a non-exported variable. I like that, actually, as it enforces the singleton pattern here.

I added a Destroy() function instead though as that could potentially come in handy down the line.

mattlord commented:
> Great work on this!
> 
> Looks good. I left a couple of optional comments.

Thank you for the review and helpful feedback! I've implemented changes based on your comments here: https://github.com/vitessio/vitess/pull/12240/commits/a2d378a15f95738808d666f79d917a46a9f9eb64

harshit-gangal commented on go/test/endtoend/vtgate/schematracker/sharded_prs/st_sharded_test.go:
```diff
@@ -156,7 +172,7 @@ func TestMain(m *testing.M) {
 		}
 
-		err = waitForVTGateAndVTTablet()
+		err = clusterInstance.WaitForVTGateAndVTTablets(5 * time.Minute)
```
this looks like too much time to wait to fail the CI run eventually.

mattlord replied:
I didn't come up with that time, I merely migrated the code to a utility function. That time came from here: https://github.com/vitessio/vitess/blame/c89fa57b6330a4a878aea82350f8525b50ad4b77/go/test/endtoend/vtgate/schematracker/sharded_prs/st_sharded_test.go#L196

harshit-gangal commented on go/vt/sidecardb/identifier_cache.go:
```diff
@@ -0,0 +1,119 @@
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// This provides a read through cache of sidecar database identifiers;
+// loading the values from an opaque backend database using a provided
+// load function.
+type IdentifierCache struct {
```
nit: go comment to start with the struct name.

mattlord replied:
OK, will update.

mattlord commented on examples/local/101_initial_cluster.sh:
```diff
@@ -19,6 +19,15 @@
 
 source ../common/env.sh
 
+# This is done here as a means to test the experimental custom
+# sidecar database name work in a wide variety of scenarios as
+# the local examples are used to test many features locally.
+# This is NOT here to indicate that you should normally use a
+# non-default (_vt) value or that it is somehow a best practice
+# to do so. In production, you should ONLY use a non-default
+# sidecar database name when it's truly needed.
+SIDECAR_DB_NAME=${SIDECAR_DB_NAME:-"__vitess_metadata"}
+
```
OK, I can change the default to `_vt` then and keep the ability to set it via the ENV variable.

mattlord commented on go/mysql/endtoend/schema_change_test.go:
```diff
@@ -84,19 +85,23 @@ func TestChangeSchemaIsNoticed(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
+			clearQuery := sqlparser.BuildParsedQuery(mysql.ClearSchemaCopy, sidecardb.GetIdentifier()).Query
+			insertQuery := sqlparser.BuildParsedQuery(mysql.InsertIntoSchemaCopy, sidecardb.GetIdentifier()).Query
+			detectQuery := sqlparser.BuildParsedQuery(mysql.DetectSchemaChange, sidecardb.GetIdentifier()).Query
```
Good point. I'll update that.

harshit-gangal commented on go/vt/sidecardb/identifier_cache.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+const (
+	ErrIdentifierCacheUninitialized  = "sidecar database identifier cache is not initialized"
+	ErrIdentifierCacheNoLoadFunction = "the load from database function has not been set"
```
looks like these are used locally and do not need to be public.

harshit-gangal commented on go/vt/sidecardb/identifier_cache.go:
```diff
@@ -0,0 +1,119 @@
+// same instance. If the cache has already been initialized then
+// it will return false indicating that your New call did not
+// create a new instance.
+func NewIdentifierCache(loadFunc func(context.Context, string) (string, error)) (*IdentifierCache, bool) {
+	created := identifierCache.CompareAndSwap(nil, &IdentifierCache{
+		load:                 loadFunc,
+		sidecarDBIdentifiers: sync.Map{},
+	})
+	return identifierCache.Load().(*IdentifierCache), created
+}
+
+func GetIdentifierCache() (*IdentifierCache, error) {
+	if identifierCache.Load() == nil {
+		return nil, vterrors.New(vtrpcpb.Code_INTERNAL, ErrIdentifierCacheUninitialized)
+	} else {
+		return identifierCache.Load().(*IdentifierCache), nil
+	}
+}
```
It would be easier at other places if we can merge these 2 methods.

mattlord replied:
Merge what two methods? GetIdentifierCache() and Get()? I don't think I like that, although it would be easier for the caller in some cases (specifically where all you want to do is Get a single value) -- in general I think it makes it a little awkward. I think you're right though in that there's no reason NOT to add a convenience function for those specific cases:
```
// GetIdentifierForKeyspace is a convenience function that returns
// the sidecar database identifier for the provided keyspace.
func GetIdentifierForKeyspace(keyspace string) (string, error) {
	cache, err := GetIdentifierCache()
	if err != nil {
		return "", err
	}
	return cache.Get(keyspace)
}
```

I will do that.

harshit-gangal replied:
My thinking was when someone calls into `GetIdentifierCache` then `IdentifierCache` gets created if not already done.

harshit-gangal replied:
This way we do not have to take care of the places where they have to be initialized first before using it.

mattlord replied:
This doesn't work, as the caller needs to provide the load function (which in practice usually means they need to have a topo server handy).

mattlord replied:
The way it is now, the user does not need to know anything about the implementation used (only the creator does). I like this abstraction and IMO it simplifies things for the user.

harshit-gangal commented on go/vt/sqlparser/utils.go:
```diff
@@ -113,3 +113,52 @@ func NormalizeAlphabetically(query string) (normalized string, err error) {
+	upd := Rewrite(in, func(cursor *Cursor) bool {
+		switch node := cursor.Node().(type) {
+		case TableName:
+			if !node.Qualifier.IsEmpty() &&
+				node.Qualifier.String() == oldQualifier.String() &&
+				node.Qualifier.String() != newQualifier.String() {
+				node.Qualifier = newQualifier
```
In the start of the function, it is already checked that olddb and newdb are not same. So, `node.Qualifier.String() != newQualifier.String()` check is not needed.

harshit-gangal commented on go/vt/sqlparser/utils.go:
```diff
@@ -113,3 +113,52 @@ func NormalizeAlphabetically(query string) (normalized string, err error) {
+				cursor.Replace(node)
+				modified = true
+			}
+		case *ShowBasic: // for things like 'show tables from _vt'
+			if !node.DbName.IsEmpty() &&
+				node.DbName.String() == oldQualifier.String() &&
+				node.DbName.String() != newQualifier.String() {
+				node.DbName = newQualifier
```
similarly here

harshit-gangal commented on go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:
```diff
@@ -643,7 +644,8 @@ func (vs *vstreamer) buildJournalPlan(id uint64, tm *mysql.TableMap) error {
 	}
 	defer conn.Close()
-	qr, err := conn.ExecuteFetch("select * from _vt.resharding_journal where 1 != 1", 1, true)
+	qr, err := conn.ExecuteFetch(fmt.Sprintf("select * from %s.resharding_journal where 1 != 1",
+		sidecardb.GetIdentifier()), 1, true)
```
should this use `sqlparser.BuildParsedQuery`

mattlord replied:
Good catch! I somehow missed this one.

harshit-gangal commented on go/vt/vttablet/tabletserver/vstreamer/rowstreamer.go:
```diff
@@ -119,7 +119,7 @@ func (rs *rowStreamer) Stream() error {
 	}
 	defer conn.Close()
-	if _, err := conn.ExecuteFetch("set names binary", 1, false); err != nil {
+	if _, err := conn.ExecuteFetch("set names 'binary'", 1, false); err != nil {
```
why is this change needed?

harshit-gangal replied:
then, its clearly a parsing bug. :(

mattlord replied:
That may be, but the "correct" syntax is to use quotes anyway so I did not find it relevant to my PR and instead used the quotes.

https://dev.mysql.com/doc/refman/8.0/en/set-names.html

harshit-gangal commented on go/vt/vttablet/tabletserver/schema/engine.go:
```diff
@@ -141,7 +143,7 @@ func (se *Engine) syncSidecarDB(ctx context.Context, conn *dbconnpool.DBConnecti
 	var exec sidecardb.Exec = func(ctx context.Context, query string, maxRows int, useDB bool) (*sqltypes.Result, error) {
 		if useDB {
-			_, err := conn.ExecuteFetch(sidecardb.UseSidecarDatabaseQuery, maxRows, false)
+			_, err := conn.ExecuteFetch(fmt.Sprintf("use %s", sidecardb.GetIdentifier()), maxRows, false)
 			if err != nil {
```
this might use `sqlparser.BuildParsedQuery`

harshit-gangal commented on go/vt/vttablet/tabletmanager/tm_init.go:
```diff
@@ -486,6 +487,33 @@ func (tm *TabletManager) createKeyspaceShard(ctx context.Context) (*topo.ShardIn
+	// update the keyspace record to the default.
+	if ks.SidecarDbName == "" {
+		ks.SidecarDbName = sidecardb.DefaultName
+		ctx, unlock, lockErr := tm.TopoServer.LockKeyspace(ctx, tablet.Keyspace, "Setting sidecar database name")
+		if lockErr != nil {
+			return nil, vterrors.Wrap(lockErr, "createKeyspaceShard: cannot GetOrCreateShard shard")
+		}
+		err = tm.TopoServer.UpdateKeyspace(ctx, ks)
+		unlock(&lockErr)
+		if err != nil {
+			return nil, vterrors.Wrap(err, "createKeyspaceShard: cannot GetOrCreateShard shard")
+		}
+		if lockErr != nil {
+			return nil, vterrors.Wrap(lockErr, "createKeyspaceShard: cannot GetOrCreateShard shard")
+		}
```
if there is lock error we will not bring up the tablet.
lock error can be multiple tablets trying to update the same thing here and lock timeout.
I think with lock timeout we should add a retry.


what I am trying to say here is. Some thoughts should be when multiple tablets come at the same time, will this cause any issue or not

mattlord replied:
If you have a non-serving keyspace, a keyspace with no tablets, and then you simultaneously bring up 10 tablets let's say... it's possible. I'll work on something for this.

harshit-gangal commented on go/vt/vtgate/planbuilder/show.go:
```diff
@@ -265,7 +268,17 @@ func buildShowVMigrationsPlan(show *sqlparser.ShowBasic, vschema plancontext.VSc
+	}
+	sidecarDBID, err := sdbidc.Get(ks.Name)
+	if err != nil {
+		return nil, vterrors.Wrapf(err, "failed to read sidecar database identifier for keyspace %q from the cache",
+			ks.Name)
+	}
```
in vtgate planner or executor, we through errors with an error code
example: `vterrors.VT13001`

harshit-gangal commented on go/vt/vtgate/planbuilder/show.go:
```diff
@@ -265,7 +268,17 @@ func buildShowVMigrationsPlan(show *sqlparser.ShowBasic, vschema plancontext.VSc
+			ks.Name)
+	}
+
+	sql := fmt.Sprintf("SELECT * FROM %s.schema_migrations", sidecarDBID)
```
should use the sqlparser method here as well

harshit-gangal reviewed (commented):
some nits, mostly looks good to me.

mattlord commented:
> some nits, mostly looks good to me.

Thanks for the helpful comments @harshit-gangal ! I believe that I've addressed all of them here: https://github.com/vitessio/vitess/pull/12240/commits/da4ee794ee600368882c138b10ba371233488c64
